---
title: "Introduction to the Cerebro interface"
author: "Roman Hillje"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    number_sections: true
    self_contained: yes
    mode: selfcontained
vignette: >
  %\VignetteIndexEntry{Introduction to the Cerebro interface}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
resource_files:
  - Cerebro_interface_files/2019-09-23_01.png
  - Cerebro_interface_files/2019-09-23_02_01.png
  - Cerebro_interface_files/2019-09-23_02_02.png
  - Cerebro_interface_files/2019-09-23_03.png
  - Cerebro_interface_files/2019-09-23_04.png
  - Cerebro_interface_files/2019-09-23_05.png
  - Cerebro_interface_files/2019-09-23_06.png
  - Cerebro_interface_files/2019-09-23_07.png
  - Cerebro_interface_files/2019-09-23_08.png
  - Cerebro_interface_files/2019-09-23_09.png
  - Cerebro_interface_files/2019-09-23_10.png
  - Cerebro_interface_files/2019-09-23_11.png
  - Cerebro_interface_files/2019-09-23_12.png
  - Cerebro_interface_files/2019-09-23_13.png
  - Cerebro_interface_files/2019-09-23_14.png
  - Cerebro_interface_files/2019-09-23_15.png
  - Cerebro_interface_files/2019-09-23_16.png
  - Cerebro_interface_files/2019-09-23_17.png
  - Cerebro_interface_files/2019-09-23_18.png
  - Cerebro_interface_files/2019-09-23_19.png
---

# Introduction to Cerebro

This article is meant to give an introduction to Cerebro for people who don't have any bioinformatic skills.
Maybe you are an experimental biologists who performed a scRNA-seq experiment and the computational biologists who analysed the data for you gave you a file and told you to load it into Cerebro.
And, maybe, he/she didn't tell you what Cerebro is and what it can be used for.
If this sounds like you, then this article should bring you up to speed.

To get yourself familiar with Cerebro, you don't need a real data file yet.
The application comes with a small data set pre-loaded which will allow you to get an idea of what Cerebro will show you about your own data.
If you prefer a more authentic feeling, you can also download one of the [example data sets](https://github.com/romanhaa/Cerebro/tree/master/examples) we have prepared for illustrative purposes.

## What you need

You need two things to start exploring your data:

1) The Cerebro application. The easiest way to get Cerebro is to download the standalone executable for either Windows or macOS (whichever operating system you are using) from the [Cerebro releases](https://github.com/romanhaa/Cerebro/releases). After the download is finished, the application can be unzipped and launched.
2) A file containing your data. This is the file you get from your colleague who did the analysis (the file name should end with `.crb` or `.rds`).

## The interface

After launching Cerebro, you will see the Cerebro dashboard with all available panels on the left.
We will go through each panel and see what information is be available there.

A few general notes:

* Please note that we have placed "info" boxes in many title bars which might provide helpful information.
* Most plots are interactive. Hover over them with your cursor to get additional information.
* Plots can be exported either to a PDF file (dimensional reductions; dedicated button in the title bar) or to a file PNG file (a small button in the top right corner of the plot that looks like a camera).
* Tables can be sorted (by clicking on the respective column title). Some can also be searched through the respective search boxes, and exported to either CSV or Excel format. Those tables which can be exported also have a "Column visibility" button which allows to hide some columns (which is useful on small screens and/or when there are many columns). Please note that in some tables some columns are hidden by default and need to be activated to be seen.

### Data info

<a target="_blank" href="Cerebro_interface_files/2019-09-23_01.png"><img src="Cerebro_interface_files/2019-09-23_01.png" alt="Data info" width="100%" height="auto" /></a>

This is the panel that we will be shown to you when you launch Cerebro.
In the top input panel you can choose your data which is the file you received from your colleague or downloaded from the examples mentioned earlier.
The input file usually has the `.crb` or `.rds` extension.
After selecting the file in the dialog box, the file will be loaded.
Please be patient during this process and wait until the progress bar says "Upload complete.".
The numbers in the different tiles will now update and give you some numbers about your data sets, including the number of cells, number of samples, number of clusters, the organism, and the experiment name.
After loading the sample, there is not much more to do here so let's move on to the "Main" panel.

In this tab, you can also find preferences about WebGL and hover info. WebGL is a technology used to accelerate generation and interaction with complex graphics, such as the dimensional reduction plot with many cells. Most modern browsers should be compatible with WebGL, but if you experience problems with the dimensional reduction plots, for example cells/dots are not visible, switching off WebGL might fix that.

### Main

<a target="_blank" href="Cerebro_interface_files/2019-09-23_02_01.png"><img src="Cerebro_interface_files/2019-09-23_02_01.png" alt="t-SNE projection with cells colored by sample" width="49%" height="auto" /></a><a target="_blank" href="Cerebro_interface_files/2019-09-23_02_02.png"><img src="Cerebro_interface_files/2019-09-23_02_02.png" alt="UMAP projection with cells colored by cell cycle" width="49%" height="auto" /></a>

In the "Main" panel, you have access to all the dimensional reductions (e.g. t-SNE or UMAP) that have been generated for this data set.
Dimensional reductions can be either 2- or 3-dimensional and aim to represent the complex expression profiles of all the cells in a more interpretable way.

The plot is interactive which means you can hover over the cells with your cursor to get extra info, zoom in and out, and pan around.
You may also export the plot to a PDF image by hitting the "export to PDF" button in the title bar.
The file will the stored on your desktop.

In the input box on the left, you can select a bunch of parameters for the plot.
For example, if there are multiple dimensional reductions, you can choose which one should be shown.
Also, you can exclude entire samples or clusters from the plot.
Sometimes, data sets can become quite large which can make the plots dreadfully slow on older machines.
In those cases, consider showing only a random subset of cells by dragging the slider down from 100% to a lower value.
Next, you can choose which meta information to color the cells by.
By default, they will be colored by the sample the cells originate from, but you can change this to any meta information that is available, e.g. cluster label, cell cycle phase, etc.
The rest of the parameters are aesthetic and allow you to adjust dot size, dot opacity (transparency), and range of the X and Y scale.

### Groups

<a target="_blank" href="Cerebro_interface_files/2019-09-23_05.png"><img src="Cerebro_interface_files/2019-09-23_05.png" alt="Cluster tree" width="49%" height="auto" /></a><a target="_blank" href="Cerebro_interface_files/2019-09-23_06.png"><img src="Cerebro_interface_files/2019-09-23_06.png" alt="Number of transcripts and expressed genes by cluster" width="49%" height="auto" /></a>

The "Groups" panel gives a group-centric view of the data.
At the top of the page, you can select one of the grouping variables that have been provided when exporting the data set.

The cluster tree shows you the similarity between clusters.
This can be useful if you want to merge clusters and want to know which ones are most similar to each other.

Then, the composition of each group by other grouping variables is provided in the form of a table and plot, followed by distributions of the number of transcripts, expressed genes, and, if available, mitochondrial and ribosomal gene expression.

As before, we also see the distribution of groups into the cell cycle phases from either Seurat or Cyclone, if they have been calculated.

### Marker genes

<a target="_blank" href="Cerebro_interface_files/2019-09-23_08.png"><img src="Cerebro_interface_files/2019-09-23_08.png" alt="Table of marker genes by sample and cluster" width="100%" height="auto" /></a>

The "Marker genes" panel provides tables of - surprise - marker genes for each grouping variable.

In the case of marker genes per group, this list conceptually resembles differential expression analysis of a bulk RNA-seq experiment.
Instead, the expression profiles of groups are expected to be more homogeneous.

Marker genes should help to identify the cell type of a group, and find new markers to purify that cell population.

It is important to know that in this analysis, each cell group is compared to all other cells combined.
This is important because it means marker genes don't have to be exclusive to one cell group.
Instead, they can be shared by multiple groups.

There are multiple parameters and cut-offs that need to be chosen during the analysis.
Check the "Marker genes" section in the "Analysis info" panel to find some details.
You can sort the genes by the average logFC, the percentage of cells that express this gene (at least 1 transcript) in the cell population of interest or the rest of cells (which it is tested against), and the adjusted p-value.
If analyzing human or mouse data, there will be an extra column indicating whether or not this gene **might** be present on the cell surface.
This information is generated by checking if the genes are associated with the gene ontology (GO) term "cell surface" (GO:0009986).
Even though this information is not 100% reliable, it might point you to genes which allow you to purify the respective cell population.

### Gene expression

<a target="_blank" href="Cerebro_interface_files/2019-09-23_10.png"><img src="Cerebro_interface_files/2019-09-23_10.png" alt="Expression of individual genes in t-SNE projection" width="49%" height="auto" /></a><a target="_blank" href="Cerebro_interface_files/2019-09-23_11.png"><img src="Cerebro_interface_files/2019-09-23_11.png" alt="Expression of individual genes by sample and cluster" width="49%" height="auto" /></a>

The "Gene expression" panel allows you to visualize the log-normalized expression of one or multiple genes.

The interface is similar to the one in the "Main" panel.
Cells are shown in the selected dimensional reduction, groups can be removed from the plot, and dot size size, opacity, as well as X and Y axis ranges can be adjusted.
The color of each cell represents the expression of the queried genes ("Genes(s)" text box in top left corner).
If inserting multiple genes, these must be posted on separate lines or separated through a comma (`,`), a space (` `), or semicolon (`;`).
The colors are updated after you press `ENTER` or `SPACE`.

In addition to the dimensional reduction plot shown in the "Main" panel, you may also choose whether cells should be plotted in random order or if you prefer to have cells with highest expression to be on top of the others.

Below the dimensional reduction you can see which genes are currently shown in the plot and which genes were not found (not present in the data set).

Further down you will find violin + box plots showing the distribution of expression of the selected genes grouped by the selected grouping variable.

The last plot of this panel shows the average expression of the 50 most expressed genes (of the ones specified by the user) across all cells to quickly spot which genes influence the color scale the most.

### Gene ID conversion

<a target="_blank" href="Cerebro_interface_files/2019-09-23_17.png"><img src="Cerebro_interface_files/2019-09-23_17.png" alt="Gene ID conversion table" width="100%" height="auto" /></a>

The "Gene ID conversion" panel provides tables which allow to convert gene IDs and names.

For the moment, only tables for mouse and human genes are available.
The tables include GENCODE identifier, ENSEMBL identifier, HAVANA identifier, gene symbol and gene type and have been derived from the GENCODE annotation version M16 (mouse) and version 27 (human).

### Analysis info

<a target="_blank" href="Cerebro_interface_files/2019-09-23_18.png"><img src="Cerebro_interface_files/2019-09-23_18.png" alt="Information about analysis" width="100%" height="auto" /></a>

The "Analysis info" panel gives you an overview of the parameters which were used during the analysis (as long as they were recorded).

Shown can be:

* General information.
* Parameters used during pre-processing (cell/gene filtering, cluster resolution, etc.).
* Gene lists used for cell cycle analysis and calculation of the percentage of mitochondrial and ribosomal gene expression.
* Parameters used for marker gene detection.
* Technical information.

### About

<a target="_blank" href="Cerebro_interface_files/2019-09-23_19.png"><img src="Cerebro_interface_files/2019-09-23_19.png" alt="About Cerebro" width="100%" height="auto" /></a>

The "About" panel provides information about Cerebro and the author.

## Changes in cerebroAppLite

Starting with version 1.4.0, cerebroAppLite is a lightweight fork of the original cerebroApp package. The following features have been removed to maintain a streamlined interface:

- Enriched pathways
- Extra material
- Most expressed genes
- Trajectories
- Gene set expression

The interface has also been updated with the following changes:

- "Load Data" tab renamed to "Data info"
- "Overview" tab renamed to "Main"
- WebGL and hover info preferences are now shown in the "Data info" tab
- Sample information boxes are more colorful
- Different icons for tabs "Data info", "Main", "Groups" and "Marker Groups"

Starting with version 1.5.0, cerebroAppLite is compatible with Seurat v5, especially with BPCells Matrix. This allows for more efficient handling of large datasets through the `expression_matrix_BPCells` parameter in `launchCerebro()`.

Starting with version 1.5.2, plot settings (size, opacity, number of cells to show) can be different in gene expression and overview, which is useful for large datasets with slow gene expression.
